### PR TITLE
Add p-projector for time history

### PR DESCRIPTION
**Pull request by kidder**
## Proposed changes

Adds only p-projection for the time history.
Will add h-projection in a subsequent PR

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/NumericalAlgorithms/Spectral/Projection.cpp [hunk 1]
```diff
@@ -392,6 +392,29 @@ projection_matrix_parent_to_child(
...
+  static const Matrix identity{};
+  auto projection_matrices = make_array<Dim>(std::cref(identity));
+  const auto source_mesh_slices = source_mesh.slices();
+  const auto target_mesh_slices = target_mesh.slices();
+  for (size_t d = 0; d < Dim; ++d) {
+    const auto source_mesh_slice = gsl::at(source_mesh_slices, d);
+    const auto target_mesh_slice = gsl::at(target_mesh_slices, d);
+    if (source_mesh_slice == target_mesh_slice) {
+      // No projection necessary, keep matrix the identity in this dimension
+    } else if (source_mesh_slice.extents(0) <= target_mesh_slice.extents(0)) {
```
Comment (nilsvu) [inline src/NumericalAlgorithms/Spectral/Projection.cpp:407]: Might as well replace `0` -> `d`

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/NumericalAlgorithms/Spectral/Projection.cpp (hunk 1 above)
Comment (kidder) [reply src/NumericalAlgorithms/Spectral/Projection.cpp:407]: it's a one dimensional Mesh...

Comment (nilsvu) [reply src/NumericalAlgorithms/Spectral/Projection.cpp:407]: Oh this is the sliced mesh 👍

**Review by nilsvu — DISMISSED** [review]

File: tests/Unit/Time/Test_History.cpp [hunk 2]
```diff
@@ -795,6 +796,29 @@ void test_history_assertions() {
...
+void test_history_output() {
+  TimeSteppers::History<double> history(2);
+  const Slab slab(0.0, 1.0);
+  history.insert(TimeStepId(true, 0, slab.start()), 0.0, 1.0);
+  const auto step_time = history.back().time_step_id.step_time();
+  const auto step_size = slab.duration() / 4;
+  history.insert(TimeStepId(true, 0, step_time, 1, step_size,
+                            (step_time + slab.duration() / 4).value()),
+                 4.0, 40.0);
+  std::string expected_output = MakeString{}
```
Comment (wthrowe) [inline tests/Unit/Time/Test_History.cpp]: [optional] This could just be a string literal; no need for `MakeString`.  Also could be const.

**Review by wthrowe — COMMENTED** [review]
Only change is optional, so let me know if you want to do it.

File: tests/Unit/Time/Test_History.cpp (hunk 2 above)
Comment (kidder) [reply tests/Unit/Time/Test_History.cpp]: @wthrowe @nilsvu squashed in this change and rebased

**Review by wthrowe — APPROVED** [review]

**Review by nilsvu — APPROVED** [review]

**Pull request merged**